src/java/org/apache/cassandra/service/StorageServiceMBean.java

```diff
@@ -723,4 +731,50 @@
     /** Every read request sent to this node from another node will result in a delay of the specified value, in seconds
      */
     public void setReadDelay(int readDelay);
+
+    class BootstrapManager
+    {
+
+        private boolean allowedToBootstrap = false;
+        private final Monitor monitor = new Monitor();
+        private final Monitor.Guard isAllowedToBootstrap = getNewGuard(monitor);
+
+        public void allowToBootstrap()
+        {
+            monitor.enter();
+            try {
+                allowedToBootstrap = true;
+            } finally {
+                monitor.leave();
+            }
+        }
+
+        public void awaitBootstrappable()
+        {
+            boolean isBootstrappable;
+            try
+            {
+                isBootstrappable = monitor.enterWhen(isAllowedToBootstrap, 5, TimeUnit.MINUTES);
```

Comment:
I would propose that we just block indefinitely, rather than wait 5 minutes then fail.